### PR TITLE
adding dotnet format docs

### LINKED ISSUE #19792: Document dotnet-format CLI tool
The [dotnet-format](https://github.com/dotnet/format/) tool is missing in the documentation at https://docs.microsoft.com/en-us/dotnet/core/additional-tools/

### PR DESCRIPTION
**Pull request by jmarolf**
Fixes #19792

## Summary

Adding documentation for the dotnet cleanup command

Related PRs
- https://github.com/dotnet/sdk/pull/18814
- https://github.com/dotnet/sdk/pull/18840

### CODE REVIEW

File: docs/core/tools/dotnet-cleanup.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-cleanup.md:17]: Are `cleanup` and `[options]` swapped here?

Comment (Youssef1313) [reply docs/core/tools/dotnet-cleanup.md:17]: What does `[command]` mean here?

File: docs/core/tools/dotnet-cleanup.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-cleanup.md:24]: Does this execute third party codefixes? Could the document be clear about this whether it's a yes or no?

**Review by Youssef1313 — COMMENTED** [review]
What's the difference between this and `dotnet format`?

File: docs/core/tools/dotnet-cleanup.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-cleanup.md:4]:
```suggestion
ms.date: 07/12/2021
```

File: docs/core/tools/dotnet-cleanup.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-cleanup.md:32]: Which options are required and which are optional? For the optional ones, what is the default?

File: docs/core/tools/dotnet-cleanup.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-cleanup.md:8]:
```suggestion
**This article applies to:** ✔️ .NET 6.x SDK and later versions
```

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:46]: We usually render arguments in all caps.

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:48]: How are they delimited -- comma, semicolon?
```suggestion
  A list of relative file or folder paths to include in formatting. All files in the current directory are formatted if empty.
```

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:52]: How is the list delimited?  
Default is missing.

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:64]:
```suggestion
  Set the verbosity level. Allowed values are `q[uiet]`, `m[inimal]`, `n[ormal]`, `d[etailed]`, and `diag[nostic]`. Default value is `m[inimal]`.
```

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:72]:
```suggestion
  Produces a JSON report in the directory specified by `<REPORT_PATH>`.
```

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:80]:
```suggestion
* Clean up all code in the solution:
```

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:86]:
```suggestion
* Clean up all code in the application project:
```

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:92]:
```suggestion
* Verify that all code is correctly formatted:
```

File: docs/core/tools/dotnet-cleanup.md
Comment (tdykstra) [inline docs/core/tools/dotnet-cleanup.md:98]:
```suggestion
* Clean up all code in the `src` and `tests` directory but not in `src/submodule-a`:
```

**Review by tdykstra — COMMENTED** [review]
Normally when a command has subcommands, one of the subcommands must be used; this is different from the others such as `dotnet nuget <command>`, the project reference commands, the package commands, and the `dotnet tool` commands.  The `<command>` (the subcommand) in this case appears to be instead a filtering option that limits the range of cleanup tasks the project handles.  With that in mind, one way to deal with this situation would be to create a "Subcommands" h2 section following or preceding the Options h2, listing each command in the same format as options in the Options section, and explaining what the command does. That seems to me preferable to following the normal format for CLI ref docs, which would be a page for each subcommand, plus in this case a page for the main command used without subcommands.

File: docs/core/tools/dotnet-format.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-format.md:24]: "here" should be avoided as link text. Also it may be better to use relative link.

File: docs/core/tools/dotnet-format.md
Comment (tdykstra) [inline docs/core/tools/dotnet-format.md:3]: Markdown formatting (backticks in this case) doesn't work in the YAML metadata block.  Also I'd follow the precedent of the VS docs which tend to prefer EditorConfig with no formatting.  Occasionally they refer to the *.editorconfig* file in Markdown, i.e., as a file name with [italic style rather than code style](https://review.docs.microsoft.com/en-us/help/contribute/text-formatting-guidelines?branch=master#italics). 

```suggestion
description: The dotnet format command formats code to match EditorConfig settings for the current directory.
```

File: docs/core/tools/dotnet-format.md
Comment (tdykstra) [inline docs/core/tools/dotnet-format.md:124]: "Clean up" is left over from when the command name was cleanup.

```suggestion
* Format all code in the solution:
```

File: docs/core/tools/dotnet-format.md
Comment (tdykstra) [inline docs/core/tools/dotnet-format.md:142]: Italics for file and folder names

```suggestion
* Clean up all code in the *src* and *tests* directory but not in *src/submodule-a*:
```

**Review by tdykstra — COMMENTED** [review]
Looks good, I left a few more minor style suggestions.

File: docs/core/tools/dotnet-format.md
Comment (jmarolf) [reply docs/core/tools/dotnet-format.md:42]: It include all "higher" levels above it so `info` implies `warn` and `error`

**Review by tdykstra — COMMENTED** [review]
The build reports that a few URLs still need to be changed to Markdown file references.

Line 38: [Suggestion] Absolute link 'https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/#index-of-rules' will be broken in isolated environments. Replace with a relative link.
Line 110: [Suggestion] Absolute link 'https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/#index-of-rules' will be broken in isolated environments. Replace with a relative link.
Line 116: [Suggestion] Absolute link 'https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/#index-of-rules' will be broken in isolated environments. Replace with a relative link.

And we need to get it into the TOC -- between dotnet clean and dotnet help -- https://github.com/dotnet/docs/blob/d398e103dbce6842d0bd775f3374e597bc95ad6b/docs/fundamentals/toc.yml#L256-L259

When do you want to merge -- Is it going to be available  in Preview 7?

**Comment by jmarolf** [comment]
> The build reports that a few URLs still need to be changed to Markdown file references.

what is the correct way to do this? relative file paths failed the build for me.

> And we need to get it into the TOC -- between dotnet clean and dotnet help --

👍 will do

> Is it going to be available in Preview 7?

yep

**Comment by tdykstra** [comment]
Try ../../fundamentals/code-analysis/quality-rules/index.md#index-of-rules

**Comment by tdykstra** [comment]
My original change suggestion for the URL failed because I inferred the Markdown file name from the URL, didn't realize it was an index.md file.

**Comment by jmarolf** [comment]
> My original change suggestion for the URL failed because I inferred the Markdown file name from the URL, didn't realize it was an index.md file.

thanks! sorry for being obtuse here.

**Comment by tdykstra** [comment]
np, that was my fault, I had to go to VS Code's link insertion tool to realize what I was missing.

File: docs/core/tools/dotnet-format.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-format.md:88]:
```suggestion
The `dotnet format whitespace` subcommand will only run formatting rules associated with whitespace formatting. For a complete list of possible formatting options that you can specify in your *.editorconfig* file, see the [whitespace formatting documentation](/visualstudio/ide/reference/options-text-editor-csharp-formatting).
```

Comment (Youssef1313) [reply docs/core/tools/dotnet-format.md]: I personally don't find the linked documentation adding anything. It shows a screenshot from the Visual Studio options, rather than actual `.editorconfig` entries for whitespace formatting.

@gewarren, is EditorConfig formatting options documented somewhere?

File: docs/core/tools/dotnet-format.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-format.md:96]: Unclear to me why the link says "code style" but links to a specific category of code style rules? for example, "[unnecessary code](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/unnecessary-code-rules)" rules?

File: docs/core/tools/dotnet-format.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-format.md:116]: Did this mean to say "code quality" instead of "code style"?

```suggestion
  A space separated list of diagnostic ids to use as a filter when fixing code quality or 3rd party issues. Default value is whichever ids are listed in the `editorconfig` file. For a list of built-in analyzer rule ids that you can specify see the list of ids see the [documentation of code-analysis quality rules](../../fundamentals/code-analysis/quality-rules/index.md).
```

**Comment by tdykstra** [comment]
I'm going to disable auto-merge before approving so we can wait until Preview 7 is released.

**Review by tdykstra — APPROVED** [review]

File: docs/core/tools/dotnet-format.md
Comment (jmarolf) [reply docs/core/tools/dotnet-format.md]: linked to index

File: docs/core/tools/dotnet-format.md
Comment (gewarren) [inline docs/core/tools/dotnet-format.md:112]:
```suggestion
#### Options
```

File: docs/core/tools/dotnet-format.md
Comment (gewarren) [inline docs/core/tools/dotnet-format.md:116]:
```suggestion
  A space-separated list of diagnostic IDs to use as a filter when fixing code quality or 3rd party issues. The default value is whichever IDs are listed in the `editorconfig` file. For a list of built-in analyzer rule IDs that you can specify, see [code-analysis quality rules](../../fundamentals/code-analysis/quality-rules/index.md).
```

File: docs/core/tools/dotnet-format.md
Comment (gewarren) [inline docs/core/tools/dotnet-format.md:130]:
```suggestion
* Format all code in the application project:
```

File: docs/core/tools/dotnet-format.md
Comment (gewarren) [inline docs/core/tools/dotnet-format.md:142]:
```suggestion
* Format all code in the *src* and *tests* directory but not in *src/submodule-a*:
```

**Review by gewarren — COMMENTED** [review]
I left a couple suggestions. `id` should be changed to `ID` throughout.

File: docs/core/tools/dotnet-format.md
Comment (gewarren) [reply docs/core/tools/dotnet-format.md]: I agree that's not a useful page to link to. What about this one? https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/formatting-rules#c-formatting-rules

File: docs/core/tools/dotnet-format.md
Comment (Youssef1313) [inline docs/core/tools/dotnet-format.md:84]: What's the difference between this command and `dotnet format --fix-whitespace`? Might worth adding a note about that.

**Comment by BillWagner** [comment]
ping @gewarren @tdykstra 

What still needs to be done on this PR?

**Comment by tdykstra** [comment]
Waiting for Preview 7 release.

**Comment by tdykstra** [comment]
@jmarolf -- we also have this issue for `dotnet format` -- https://github.com/dotnet/docs/issues/19792 -- which suggests it belongs under Additional tools.  Will this be included as one of the core CLI commands or will it need to be installed as an additional tool?

**Comment by jmarolf** [comment]
> Will this be included as one of the core CLI commands or will it need to be installed as an additional tool?

this is a core command, there is no need to document it as a tool. The preferred way that developers should interact with it going forward is as a core command

**Pull request merged**
